src/main/java/com/pinterest/secor/io/impl/FlexibleDelimitedFileReaderWriterFactory.java

```diff
@@ -0,0 +1,166 @@
+*/
+package com.pinterest.secor.io.impl;
+
+import java.io.BufferedInputStream;
+import java.io.BufferedOutputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.InputStream;
+
+import com.pinterest.secor.io.FileReader;
+import com.pinterest.secor.io.FileReaderWriterFactory;
+import com.pinterest.secor.io.FileWriter;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.io.compress.CompressionCodec;
+import org.apache.hadoop.io.compress.Compressor;
+import org.apache.hadoop.io.compress.CodecPool;
+import org.apache.hadoop.io.compress.Decompressor;
+
+import org.apache.commons.configuration.ConfigurationException;
+
+import com.google.common.io.CountingOutputStream;
+import com.pinterest.secor.common.LogFilePath;
+import com.pinterest.secor.io.KeyValue;
+import com.pinterest.secor.util.FileUtil;
+import com.pinterest.secor.common.SecorConfig;
+
+/**
+* Flexible Delimited Text File Reader Writer with Compression
+*
+* @author Ahsan Nabi Dar (ahsan@wego.com)
+*/
+public class FlexibleDelimitedFileReaderWriterFactory implements FileReaderWriterFactory {
+
+  @Override
+  public FileReader BuildFileReader(LogFilePath logFilePath, CompressionCodec codec)
+  throws IllegalAccessException, IOException, InstantiationException {
+    return new FlexibleDelimitedFileReader(logFilePath, codec);
+  }
+
+  @Override
+  public FileWriter BuildFileWriter(LogFilePath logFilePath, CompressionCodec codec) throws IOException {
+    return new FlexibleDelimitedFileWriter(logFilePath, codec);
+  }
+
+
+
+  protected class FlexibleDelimitedFileReader implements FileReader {
+    private final BufferedInputStream mReader;
+    private long mOffset;
+    private Decompressor mDecompressor = null;
+    private byte DELIMITER = delimterReader();
+
+    public FlexibleDelimitedFileReader(LogFilePath path, CompressionCodec codec) throws IOException {
+      Path fsPath = new Path(path.getLogFilePath());
+      FileSystem fs = FileUtil.getFileSystem(path.getLogFilePath());
+      InputStream inputStream = fs.open(fsPath);
+      this.mReader = (codec == null) ? new BufferedInputStream(inputStream)
+      : new BufferedInputStream(
+      codec.createInputStream(inputStream,
+      mDecompressor = CodecPool.getDecompressor(codec)));
+      this.mOffset = path.getOffset();
+    }
+
+    public byte delimterReader() {
+      byte delimiter = '\n';
+      try {
+        String readerDelimiter = SecorConfig.load().getFileReaderDelimiter();
+        if (!readerDelimiter.isEmpty()){
+          delimiter = (byte)readerDelimiter.charAt(0);
+        }
+      }catch(ConfigurationException e){}
+        return delimiter;
+      }
+
+      @Override
+      public KeyValue next() throws IOException {
+        ByteArrayOutputStream messageBuffer = new ByteArrayOutputStream();
+        int nextByte;
+        while ((nextByte = mReader.read()) != DELIMITER) {
```

Comment:
Delimiter can't be part of data otherwise no way to identify where new message start and where it ends, it needs to be unique, no need for that case